storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -801,8 +803,23 @@ private Void deleteTypeIfExists(StorageAction<Boolean, IOException> delete, Stri
         try {
             if (delete.execute())
                 LOGGER.info("Deleted {} {}.", fileType, file.getAbsolutePath());
-            else if (logIfMissing)
-                LOGGER.info("Failed to delete {} {} because it does not exist.", fileType, file.getAbsolutePath());
+            else {
+                if (logIfMissing) {
+                    LOGGER.info("Failed to delete {} {} because it does not exist.", fileType, file.getAbsolutePath());
+                }
+
+                // During alter log dir, the log segment may be moved to a new directory, so async delete may fail.
+                // Fallback to delete the file in the new directory to avoid orphan file.
+                Pattern dirPattern = Pattern.compile("^(\\S+)-(\\S+)\\.(\\S+)-future");
```

Comment:
nit: The `future` string can be replaced with a variable. Also the regex can also become a variable, to make it readable.

Reply:
Updated it. Thanks for the review.

Reply:
I think we can create a file like `LogDirUtils.java` in the future, so we don't need to define same variables like `FutureDirPattern` and `FutureDirSuffix` in `LogSegment.java` and `LocalLog.scala`. WDYT?

Reply:
Sounds good.